File: opcode/opcode.hs
module Main where

import Control.Monad (forM_)

data Opcode = BRK | JSR
    deriving (Enum, Bounded, Show)

data OpcodeInfo = OpcodeInfo {opcode :: Int, size :: Int}

opcodeInfo :: Opcode -> OpcodeInfo
opcodeInfo BRK = OpcodeInfo 0x00 1
opcodeInfo JSR = OpcodeInfo 0x20 3

main = forM_ [minBound .. maxBound] $ \opcode ->
    let OpcodeInfo code size = opcodeInfo opcode
     in putStrLn $
            show opcode ++ "=" ++ show code
                ++ " size="
                ++ show size

File: opcode/opcode.rs
#[derive(Clone, Copy, Debug)]
enum Op {
    BRK,
    JSR,
}
use Op::*;

struct Info {
    code: u32,
    size: u8,
}

fn info(op: Op) -> Info {
    match op {
        BRK => Info {
            code: 0x00,
            size: 1,
        },
        JSR => Info {
            code: 0x20,
            size: 3,
        },
    }
}

fn main() {
    for op in vec![BRK, JSR] {
        let i = info(op);
        println!("{:?}= {} (size {})", op, i.code, i.size)
    }
}
